examples/speaker_diarization/speaker_diarization/utils.py

```diff
@@ -0,0 +1,244 @@
+
+
+def compute_intersection_length(A: List[Tuple[float, float]], B: List[Tuple[float, float]]) -> float:
+    # borrowed from https://github.com/wq2012/SimpleDER/blob/master/simpleder/der.py
+    max_start = max(A[1], B[1])
+    min_end = min(A[2], B[2])
+    return max(0.0, min_end - max_start)
+
+
+def build_cost_matrix(ref, inf) -> dict:
+    # borrowed from https://github.com/wq2012/SimpleDER/blob/master/simpleder/der.py
+    ref_index = build_speaker_index(ref)
+    inf_index = build_speaker_index(inf)
+    cost_matrix = np.zeros((len(ref_index), len(inf_index)))
+    for ref_element in ref:
+        for hyp_element in inf:
+            i = ref_index[ref_element[0]]
+            j = inf_index[hyp_element[0]]
+            cost_matrix[i, j] += compute_intersection_length(
+                ref_element,
+                hyp_element,
+            )
+    return cost_matrix
+
+
+def realign_labels(ref_df: pd.DataFrame, inf_df: pd.DataFrame):
+    """
+    Aligns speaker labels using linear sum optimiztion
+    """
+    ref = [(row.speaker, row.starttime, row.endtime) for i, row in ref_df.iterrows()]
+    inf = [(row.speaker, row.starttime, row.endtime) for i, row in inf_df.iterrows()]
+    cost_matrix = build_cost_matrix(ref, inf)
+    row_index, col_index = optimize.linear_sum_assignment(-cost_matrix)
+
+    ref_dict = build_speaker_index(ref)
+    ref_dict_inv = {v: k for k, v in ref_dict.items()}
+    inf_dict = build_speaker_index(inf)
+    inf_dict_inv = {v: k for k, v in inf_dict.items()}
+
+    mapping = {}
+    for i in range(len(col_index)):
+        mapping[inf_dict_inv[col_index[i]]] = ref_dict_inv[row_index[i]]
+
+    inf_df["speaker"] = inf_df["speaker"].apply(lambda x: f"NA_{int(x)}" if x not in mapping.keys() else mapping[x])
+
+
+def generate_annotation(segments: List[TimeSegment]) -> Annotation:
+    """
+    Generates pyannote Annotation objects for calculating metrics
+    """
+    annotation = Annotation()
+    for segment in segments:
+        annotation[Segment(segment.start, segment.end)] = segment.group
+
+    return annotation
+
+
+def preprocess_text(segments: List[TimeSegment]) -> str:
+    """
+    Preprocess text by removing punctuation and combining transcriptions.
+    """
+    text = " ".join([segment.label for segment in segments])
+    text = re.sub(r"[^\w\s]", "", text.lower())
+    return text
+
+
+def create_non_overlapping_segments(
+    transcription: List[LabeledTimeSegment],
+    identity=None,
+) -> List[Tuple[float, float]]:
+    """
+    Takes in a list of TimeSegments (overlapping or not) and returns a non-overlapping segment list.
+    """
+    res = []  # [(start, end), ...]
+    transcription = sorted(transcription, key=lambda x: x.start)
+
+    for t in transcription:
+        if identity is not None and t.group != identity:
+            continue
+
+        start_time = t.start
+        end_time = t.end
+
+        start_idx = -1
+        end_idx = -1
+        for i, (s, e) in enumerate(res):
```

Comment:
I think this part can be optimized further by not checking against the entire `res` if compute time is an issue.